### PR TITLE
Implement Hadamard Gaussian noise

### PR DESCRIPTION
**Pull request by TobyBoyne**
Addresses #877, #2416. Implements noise that is homoskedastic in inputs, but task-dependent.

Existing multi-task noises (https://docs.gpytorch.ai/en/latest/likelihoods.html#multi-dimensional-likelihoods) only work where all tasks are observed for each input. In the 'Hadamard' setting, where each input corresponds to exactly one task, there is no existing implementation that supports a different noise for each task.

See the updated `Hadamard Multitask GP Regression` for usage.

Unit tests, documentation, type hinting will all also be provided soon!

### CODE REVIEW

**Comment by TobyBoyne** [comment]
Two areas that likely need improving:
- The implementation of `_shaped_noise_covar` probably needs some work. I'm not sure if there is some `linear_operator` magic to make this neater?
- The syntax can be a little questionable. For example, I don't like the fact that the task indexes need to be in a list in `mll(output, full_train_y, [full_train_i])`. It is also not obvious that the implementation requires the task indexes to be passed as the last argument to the model, since they are accessed by `params[0][-1]`. At least some type checking is required here?

Also, I should credit @jpfolch for help with this implementation!

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+from ..distributions import base_distributions, MultivariateNormal
+from ..likelihoods import _GaussianLikelihoodBase
+from .noise_models import MultitaskHomoskedasticNoise
+
+
+class HadamardGaussianLikelihood(_GaussianLikelihoodBase):
+    r"""
+    Likelihood for input-wise homo-skedastic noise, and task-wise
+    hetero-skedastic, i.e. we learn a different (constant) noise level for each fidelity.
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:17]:
```suggestion
    Likelihood for input-wise homoskedastic and task-wise heteroskedastic noise,
    i.e. we learn a different (constant) noise level for each fidelity.
```

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    Args:
+        num_of_tasks : number of tasks in the multi output GP
+        noise_prior : any prior you want to put on the noise
+        noise_constraint : constraint to put on the noise
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:22]:
```suggestion
        num_of_tasks: Number of tasks in the multi-output GP.
        noise_prior: Prior for the noise.
        noise_constraint: Constraint on the noise value.
```

Comment (Balandat) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:22]: `batch_shape` is missing here

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    """
+
+    def __init__(
+        self,
+        num_tasks,
+        noise_prior=None,
+        noise_constraint=None,
+        batch_shape=torch.Size(),
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:30]: can you add type annotations here?

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    Args:
+        num_of_tasks : number of tasks in the multi output GP
+        noise_prior : any prior you want to put on the noise
+        noise_constraint : constraint to put on the noise
+    """
+
+    def __init__(
+        self,
+        num_tasks,
+        noise_prior=None,
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:28]: So we can only define a single prior for all tasks here - I think that's ok but probably worth mentioning this limitation in the docstring.

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    def raw_noise(self) -> torch.Tensor:
+        return self.noise_covar.raw_noise
+
+    @raw_noise.setter
+    def raw_noise(self, value: torch.Tensor) -> None:
+        self.noise_covar.initialize(raw_noise=value)
+
+    def _shaped_noise_covar(self, base_shape: torch.Size, *params: Any, **kwargs: Any):
+        # params contains task indexes
+        task_idxs = params[0][-1]
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:60]: Language nit
```suggestion
        task_idcs = params[0][-1]
```

Comment (Balandat) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:60]: As you said, some runtime typechecking to make sure these are indeed the task indices would probably be good here.

Comment (Balandat) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:60]: Also, if you can introduce some comments with the shapes of the intermediate tensors / matrices in this function that would be helpful.

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        mask = DiagLinearOperator(diag)
+        return (noise_base_covar_matrix @ mask).sum(dim=-3)
+
+    def forward(
+        self,
+        function_samples: torch.Tensor,
+        *params: Any,
+        **kwargs: Any,
+    ) -> base_distributions.Normal:
+        noise = self._shaped_noise_covar(function_samples.shape, *params, **kwargs).diag()
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:74]: It's a bit unfortunate that IIUC `self._shaped_noise_covar` always returns the full covariance but that you're calling `diag()` on it here - ideally we can have a way to just compute the diagonal here.

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        return base_distributions.Normal(function_samples, noise.sqrt())
+
+    def marginal(self, function_dist: MultivariateNormal, *params: Any, **kwargs: Any) -> MultivariateNormal:
+        mean, covar = function_dist.mean, function_dist.lazy_covariance_matrix
+        noise_covar = self._shaped_noise_covar(mean.shape, *params, **kwargs).squeeze(0)
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:79]: Why is the `.squeeze(0)` necessary here? Does it work for general batch shapes?

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    def _shaped_noise_covar(self, base_shape: torch.Size, *params: Any, **kwargs: Any):
+        # params contains task indexes
+        task_idxs = params[0][-1]
+        noise_base_covar_matrix = self.noise_covar(*params, shape=base_shape, **kwargs)
+
+        all_tasks = torch.arange(self.num_tasks)[:, None]
+        diag = torch.eq(all_tasks, task_idxs.mT)
+        mask = DiagLinearOperator(diag)
+        return (noise_base_covar_matrix @ mask).sum(dim=-3)
```
Comment (Balandat) [inline gpytorch/likelihoods/hadamard_gaussian_likelihood.py:66]: Could you use a [`MaskedLinearOperator`](https://github.com/cornellius-gp/linear_operator/blob/2789d185f9d8b7f6e93bc6c350c031d5ad82a9fd/linear_operator/operators/masked_linear_operator.py#L12) here? Not sure if that would necessarily be better (seems like it wouldn't help with the `diag` comment below b/c of [this](https://github.com/cornellius-gp/linear_operator/blob/2789d185f9d8b7f6e93bc6c350c031d5ad82a9fd/linear_operator/operators/masked_linear_operator.py#L79)).

File: examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.ipynb [hunk 9]
```diff
@@ -29,7 +29,22 @@
    "cell_type": "code",
    "execution_count": 1,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
Comment (Balandat) [inline examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.ipynb:32]: ideally we could avoid emitting this unrelated & unnecessary warning

File: examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.ipynb [hunk 10]
```diff
@@ -108,7 +123,7 @@
     "        \n",
     "        return gpytorch.distributions.MultivariateNormal(mean_x, covar)\n",
     "\n",
-    "likelihood = gpytorch.likelihoods.GaussianLikelihood()\n",
+    "likelihood = gpytorch.likelihoods.HadamardGaussianLikelihood(num_tasks=2)\n",
```
Comment (Balandat) [inline examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.ipynb:126]: Rather than changing the tutorial to only use task-specific noises, it would be good to instead show both cases and discuss the pros and cons of both approaches in the tutorial. Ideally you could produce the output you shared in https://github.com/pytorch/botorch/issues/2765#issuecomment-2715182824

We could also try to highlight a potential failure mode where in the very low data regime andwith uniform noise levels across tasks, estimating task-specific noises results in worse performance due to the less parsimonious model.

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 6 above)
Comment (TobyBoyne) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:74]: From my understanding, if `self._shaped_noise_covar` returns a LinearOperator, then calling `.diagonal` should only calculate the diagonal (as in `_MultiTaskGaussianLikelihoodBase.forward()`). I'm not quite sure if my `_shaped_noise_covar` function does return a correct LinearOperator that shows this behaviour. I'll look into `MaskedLinearOperator` as you mentioned below.

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 4 above)
Comment (TobyBoyne) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:28]: Would it not be possible to have a multi-dimensional prior on your task noises? I had assumed that this will work by-default with the current implementation, but I will confirm this behaviour works as intended.
```python
noise_prior=LogNormalPrior(
    loc=torch.tensor([0., 1.]), 
    scale=torch.tensor([1., 1.])
)
```

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 5 above)
Comment (TobyBoyne) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py]: Is there a "canonical" way to check that these are the task indices? I will check shape and dtype, but I wasn't sure if there was a way to guarantee that `task_idcs` contains what we want.

**Comment by TobyBoyne** [comment]
Thanks for the thorough review, @Balandat! I've left a couple of comments and will update the PR soon to reflect your suggestions.

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 4 above)
Comment (Balandat) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:28]: hmm you're right, that should work with a multi-dimensional prior (but not with different types of priors on different tasks)

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 5 above)
Comment (Balandat) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py]: I don't think there is, unfortunately

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 6 above)
Comment (Balandat) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:74]: > I'm not quite sure if my _shaped_noise_covar function does return a correct LinearOperator that shows this behaviour

I was worried since you're using `.sum(dim=-3)` in  `self._shaped_noise_covar`, will that return a `LinearOperator` in your case?

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 7 above)
Comment (TobyBoyne) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:79]: The `squeeze` here is due to the shape of `MultitaskHomoskedasticNoise.forward`. It seems to add an extra dimension than I would expect, as the output is `(1, num_tasks, num_data, num_data)`. I have changed this to `.squeeze(-4)`, and moved it to where it is actually relevant for clarity.

```python
base_shape = torch.Size([40])
HomoskedasticNoise()(shape=base_shape).shape # (40, 40)
MultitaskHomoskedasticNoise(num_tasks=2)(shape=base_shape).shape # (1, 2, 40, 40)
```

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 6 above)
Comment (TobyBoyne) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:74]: I have checked, and yes - the function does return a `LinearOperator`, since `sum`ming a `LinearOperator` will return another `LinearOperator` here.

File: gpytorch/likelihoods/hadamard_gaussian_likelihood.py (hunk 8 above)
Comment (TobyBoyne) [reply gpytorch/likelihoods/hadamard_gaussian_likelihood.py:66]: I don't think it can be used here. In my code, `diag`, which is the diagonal of the mask matrix, has shape `(num_tasks, num_data)`. However, it looks like `MaskedLinearOperator` would only allow for a mask of shape `(num_data,)`.

This could feasibly be rewritten as a comprehension, summing over `num_tasks` different `MaskedLinearOperator`, but I don't think it would help with the readability of the code, which does already return a `LinearOperator`.

**Comment by TobyBoyne** [comment]
Hi @Balandat, I've pushed some changes that should address some of the concerns raised:
- Added tests to check behaviour in batched settings
- Rearranged notebook to discuss pros and cons of task-specific noise in appendix
- Added the requested type hints and shape comments

I'm not sure why the readthedocs check is failing - seems to be having trouble importing `matplotlib` just for my notebook?

Hope you have a weekend! Please let me know if there are any other changes to be made.

**Review by Balandat — COMMENTED** [review]
Thanks, code looks great. The notebook has no outputs included though? Can you re-run this and make sure that it includes the figures so that users can see the plots without having to run them themselves (and so they render on GitHub)? 

> I'm not sure why the readthedocs check is failing - seems to be having trouble importing matplotlib just for my notebook?

Not sure what this is - could be transient, let's just try again maybe it'll work on another push.

**Comment by TobyBoyne** [comment]
Hi @Balandat, were there any other changes for this PR that need to be completed before it can be merged?

**Review by Balandat — APPROVED** [review]
No, this just fell off my radar. Thanks!
